### PR TITLE
CP-49365: Implement xenserver_network_vlan provider resource

### PR DESCRIPTION
**Pull request by xueqingz**
test_result:
```
ubuntu@ubuntu-server:~/code/terraform-provider-xenserver$ make testacc
if [ -z "/home/ubuntu/go/bin" ]; then echo "GOBIN is not set" && exit 1; fi
go mod tidy
go install .
ls -l /home/ubuntu/go/bin/terraform-provider-xenserver
-rwxrwxr-x 1 ubuntu ubuntu 22787058 Jul  3 07:36 /home/ubuntu/go/bin/terraform-provider-xenserver
source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkDataSource
--- PASS: TestAccNetworkDataSource (1.52s)
=== RUN   TestAccVlanResource
--- PASS: TestAccVlanResource (4.09s)
=== RUN   TestAccNICDataSource
--- PASS: TestAccNICDataSource (0.67s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (0.67s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (0.67s)
=== RUN   TestAccNFSResource
--- PASS: TestAccNFSResource (4.74s)
=== RUN   TestAccSRResourceLocal
--- PASS: TestAccSRResourceLocal (4.15s)
=== RUN   TestAccSRResourceShared
--- PASS: TestAccSRResourceShared (7.76s)
=== RUN   TestAccVDIResource
--- PASS: TestAccVDIResource (5.46s)
=== RUN   TestAccVMDataSource
--- PASS: TestAccVMDataSource (1.34s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (1.58s)
PASS
ok      terraform-provider-xenserver/xenserver  32.662s
```

### CODE REVIEW

**Review by JinyuZhou-patty — APPROVED** [review]

File: xenserver/network_vlan_resource.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+		resp.Diagnostics.AddError(
+			"Unable to get network ref",
+			err.Error(),
+		)
+		return
+	}
+	err = vlanResourceModelUpdate(ctx, r.session, networkRef, data)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to update network_vlan resource",
```
Comment (acefei) [inline xenserver/network_vlan_resource.go:245]: replace `"Unable ` to `"unable `

File: xenserver/network_utils.go [hunk 2]
```diff
@@ -86,64 +87,237 @@ func updateNetworkRecordData(ctx context.Context, record xenapi.NetworkRecord, d
...
-	err = xenapi.Network.SetMTU(session, networkRef, int(data.MTU.ValueInt64()))
+	return pifRefs, nil
+}
+
+func getVlanCreateParams(session *xenapi.Session, data vlanResourceModel, networkRef xenapi.NetworkRef) (vlanCreateParams, error) {
+	var params vlanCreateParams
+	pifRefs, err := getPifRefsForNIC(session, data.NIC.ValueString())
 	if err != nil {
-		return errors.New("unable to update network mtu")
+		return params, errors.New(err.Error())
```
Comment (acefei) [inline xenserver/network_utils.go:198]: can replace it to `return params, err.Error()`? or `return params, err`

File: xenserver/network_vlan_resource.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+			err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *vlanResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data vlanResourceModel
```
Comment (acefei) [inline xenserver/network_vlan_resource.go:212]: suggest to replace the definition of `data` and `dataState`  to `var plan state vlanResourceModel`, which is more readable.

File: xenserver/network_utils.go [hunk 4]
```diff
@@ -86,64 +87,237 @@ func updateNetworkRecordData(ctx context.Context, record xenapi.NetworkRecord, d
...
+type vlanCreateParams struct {
+	PifRef     xenapi.PIFRef
+	NetworkRef xenapi.NetworkRef
+	Tag        int
+}
 
-	err := updateNetworkResourceModelComputed(ctx, networkRecord, data)
-	if err != nil {
-		return err
+func checkMTU(mtu int) error {
```
Comment (acefei) [inline xenserver/network_utils.go:107]: where do we find this limition reference?

Comment (xueqingz) [reply xenserver/network_utils.go:107]: It's the mtu limit in XC side. But year, let me confirm with toolstack.

File: xenserver/network_vlan_resource.go (hunk 1 above)
Comment (xueqingz) [reply xenserver/network_vlan_resource.go:245]: we should keep same with other error format.

**Review by acefei — APPROVED** [review]

File: xenserver/network_utils.go (hunk 4 above)
Comment (xueqingz) [reply xenserver/network_utils.go:107]: the mtu should > 0

**Pull request merged**